compiler/luci/pass/src/FuseHorizontalFCLayers.cpp

```diff
@@ -0,0 +1,195 @@
+    left->template at<D>(i) += right->template at<D>(i);
+  }
+}
+
+bool fuse_horizontal_fc_nodes(luci::CircleAdd *add_node)
+{
+  // Let's check left and right FC nodes
+  auto left_fc_node = dynamic_cast<luci::CircleFullyConnected *>(add_node->x());
+  auto right_fc_node = dynamic_cast<luci::CircleFullyConnected *>(add_node->y());
+
+  if (left_fc_node == nullptr or right_fc_node == nullptr)
+    return false;
+
+  if (not check_type_and_shape_equality(left_fc_node, right_fc_node))
+    return false;
+
+  if (left_fc_node->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  if (right_fc_node->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  // Let's check that FC nodes have the same input
+  if (left_fc_node->input() != right_fc_node->input())
+    return false;
+
+  // Lets check left and right FC weights: type and shape
+  auto left_fc_weights = dynamic_cast<luci::CircleConst *>(left_fc_node->weights());
+  auto right_fc_weights = dynamic_cast<luci::CircleConst *>(right_fc_node->weights());
+
+  if (left_fc_weights == nullptr or right_fc_weights == nullptr)
+    return false;
+
+  if (not check_type_and_shape_equality(left_fc_weights, right_fc_weights))
+    return false;
+
+  // Lets check left and right FC bias: type and shape
+  auto left_fc_bias = dynamic_cast<luci::CircleConst *>(left_fc_node->bias());
+  auto right_fc_bias = dynamic_cast<luci::CircleConst *>(right_fc_node->bias());
+
+  // Support only if both biases are const, or both are non-const
+  // TODO Support the case that one FC has a const bias and another FC has no bias.
+  if ((left_fc_bias == nullptr and right_fc_bias != nullptr) or
+      (left_fc_bias != nullptr and right_fc_bias == nullptr))
+  {
+    return false;
+  }
+
+  // Both left/right bias are const. Check dtype/shape.
+  if (left_fc_bias != nullptr and not check_type_and_shape_equality(left_fc_bias, right_fc_bias))
+    return false;
+
+  // Both left/right bias are non-const. Check left/right fc has no bias.
+  if (left_fc_bias == nullptr)
+  {
+    auto left_no_bias = dynamic_cast<luci::CircleOutputExclude *>(left_fc_node->bias());
+    auto right_no_bias = dynamic_cast<luci::CircleOutputExclude *>(right_fc_node->bias());
+    if (not left_no_bias or not right_no_bias)
+      return false;
+  }
+
+  // Lets create fused FC weights and bias
+  auto fused_fc_weights = luci::clone(left_fc_weights);
+  luci::add_origin(fused_fc_weights, luci::composite_origin({luci::get_origin(left_fc_weights),
+                                                             luci::get_origin(right_fc_weights)}));
+
+  luci::CircleConst *fused_fc_bias = nullptr;
+  if (left_fc_bias != nullptr)
+  {
+    fused_fc_bias = luci::clone(left_fc_bias);
+    luci::add_origin(fused_fc_bias, luci::composite_origin({luci::get_origin(left_fc_bias),
+                                                            luci::get_origin(right_fc_bias)}));
+  }
+
+  switch (left_fc_weights->dtype())
```

Comment:
Please check dtype before `luci::clone` is called. `luci::clone` creates a new node, which changes a graph.

Reply:
Done